### PR TITLE
Adding tree view for local library (plus other fixes)

### LINKED ISSUE #568: Scrolling issue with new simple list in local library
- **Nuclear version**: 0.5.0 commit 331521a7ba41c898ced62e97365b60295a05cf6e
- **OS**: Windows 10 Pro x64 1909
- **NodeJS version**: 12.12.0
---
**Description of the issue**: Since the new album layout has been added to the local library there have been some scrolling issues with the simple list. The scrollbar constantly jitters around and completely resets if dragged too low.
Here is a short video showcasing the issue:
[https://streamable.com/357vq](https://streamable.com/357vq)

### LINKED ISSUE #685: [feature request] Tree-view for local library
Most of the long-term music players I've used in the past have had a way to display one's local library within a tree view, matching the folder/file structure.

For example, in foobar2000:

It would be nice to have an option to display one's local library like this in Nuclear as well. By displaying the actual folder/file structure, it allows you to categorize and subcategorize by your own criteria (for example, the `@Games` folder above), without it all just getting "flattened" as in the regular track/album views. Those views are good for many people, but some prefer the full control of having the raw directory tree represented.

I'm willing to work on this (as I also need it for [my plugin](https://github.com/Venryx/nuclear-vplugin) anyway), but would like to start the discussion here first to know if/how @nukeop would like me to approach it.

### PR DESCRIPTION
**Pull request by Venryx**
Tree-view-related changes:
* Added folder-tree view to local library page. (resolves #685)
* Greatly improved ui responsiveness, when track was playing, and the folder-tree view was open. (by using react-hooks useMemo in LibraryView component to prevent the "tracks" prop from changing, which was causing a new treeConfig to get sent to react-redux-grid each tick, causing it to rerender its internals -- specifically, deep-comparing the data tree)
* Added new `local.expandedFolders` store field, which remembers the expanded rows of the local-library folder tree-view. (and gets persisted to disk on program close)
* As required for persistence of `local.expandedFolders`, added support for nested fields to syncStorage enhancer.

General changes: (mostly fixes)
* Fixed that local tracks would not play, due to incorrect stream selection.
* Fixed that local-library folder-paths were not getting normalized (eg. from `\\` to `/`, on Windows) on initial selection. (fixes inconsistency, since gets normalized after closing and relaunching -- and consistency needed by new track-population code)
* React-list (for simple-list view of local library) was correctly setting the `height` value for the table, but for some reason it was being ignored by the browser renderer with the default `display: block` style. Switched the display-type to `table` to fix this. (resolves #568)
* Fixed error when opening the "Add local folder" dialog, then pressing "Cancel".
* Renamed NuclearBrutMeta.pos to NuclearBrutMeta.position, to match with all other "position" fields in the project. Also fixes React warning in Album view of local library, which was looking for "position" field. (TypeScript integration throughout the codebase should really help with this sort of thing.)
* Fixed that AlbumPreview component was still having React warnings, due to `track.position` apparently having clashes / not always being unique. (now just uses index as key)
* Corrected ui/Card.PropTypes.image to match the actual data. (thus solving React warning)
* Fixed that FavoriteTracksView was using minHeight instead of height, causing "no favorite tracks" text to not be centered. (and for scrolling, improved styling by having scrolling occur in table specifically, not for entire container, as that would cause darker background to not scroll with)
* Fixed PropType definition for TrackRow.track.artist, in ui/TrackRow/index.js. (fixing warning)
* Fixed React warning in ui/TrackRow, about `withAddToDownloads` not being valid attribute for "tr" element.

Screenshot of new folder-tree view:

### CODE REVIEW

**Comment by Venryx** [comment]
Hmm, the CI build is failing: https://travis-ci.org/github/nukeop/nuclear/jobs/684086426

However, on reading the logs, it appears its caused by this:
```
⨯ cannot resolve https://github.com/electron-userland/electron-builder-binaries/releases/download/nsis-resources-3.4.1/nsis-resources-3.4.1.7z: status code 429
```

I'm assuming it's just a network error? If so, how to restart the CI build process?

**Comment by nukeop** [comment]
I restarted your build on Travis, let's see how it goes this time.

**Comment by nukeop** [comment]
Nice, it passed.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+          id: path,
+          path,
+          name: path.split('/').slice(-1)[0],
+          children: []
+        };
+  
+        if (localFolders.includes(path)) {
+          newEntry.parentId = -1;
+        } else if (path.includes('/')) {
+          /* if (!localFolders.find(folderPath => path.startsWith(folderPath))) {
```
Comment (nukeop) [inline packages/app/app/components/LibraryView/LibraryFolderTree/index.js:27]: Please remove the commented out code

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        const newEntry = {
+          id: path,
+          path,
+          name: path.split('/').slice(-1)[0],
+          children: []
+        };
+  
+        if (localFolders.includes(path)) {
+          newEntry.parentId = -1;
+        } else if (path.includes('/')) {
```
Comment (nukeop) [inline packages/app/app/components/LibraryView/LibraryFolderTree/index.js:26]: I think that the best course of action for platform compatibility would be to use `path.sep` instead of `/`

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  );
+};
+
+LibraryFolderTree.propTypes = {
+  tracks: PropTypes.array,
+  localFolders: PropTypes.arrayOf(PropTypes.string),
+  expandedFolders: PropTypes.arrayOf(PropTypes.string)
+};
+
+export default compose(
```
Comment (nukeop) [inline packages/app/app/components/LibraryView/LibraryFolderTree/index.js:164]: Let's forget about hocs in this component and use `useTranslation` instead, compose is also unneeded.

File: packages/app/app/components/LibraryView/LibraryFolderTree/styles.scss [hunk 4]
```diff
@@ -0,0 +1,8 @@
+$header-background-color: transparent;
+$background-color: transparent;
+//$table-prefix: BaseTable !default;
+
+$border: none;
+$row-hovered-background-color: hsla(0, 0%, 100%, .2);
+
+@import '~react-base-table/es/_BaseTable.scss';
```
Comment (nukeop) [inline packages/app/app/components/LibraryView/LibraryFolderTree/styles.scss:8]: Can this be moved to the top of the file?

File: packages/app/app/components/LibraryView/LibrarySimpleList/index.js [hunk 5]
```diff
@@ -23,6 +23,12 @@ const LibrarySimpleList = ({
   t
 }) => (
   <ReactList
+    ref={comp => {
```
Comment (nukeop) [inline packages/app/app/components/LibraryView/LibrarySimpleList/index.js:26]: Does this have to use a ref? Why not use css?

File: packages/app/app/components/LibraryView/index.js [hunk 6]
```diff
@@ -1,35 +1,56 @@
-import React, { useCallback } from 'react';
-import PropTypes from 'prop-types';
+import React, { useCallback, useMemo } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
```
Comment (nukeop) [inline packages/app/app/components/LibraryView/index.js:2]: State subscriptions and action creator bindings should be kept in the corresponding container, and passed down as needed. Here they're split between the container and the component.

File: packages/app/app/store/enhancers/syncStorage.js [hunk 7]
```diff
@@ -8,11 +8,16 @@ export default function(paths) {
       initialState = undefined;
     }
 
-    let persistedState, finalInitialState;
+    // retrieve initial states from reducers, to apply prior to merging electron-store persisted-data
```
Comment (nukeop) [inline packages/app/app/store/enhancers/syncStorage.js:11]: Why is this needed?

We already have a mechanism for applying default values to settings that weren't touched, unless this is needed for a different reason.

**Comment by nukeop** [comment]
Thanks for fixing so many things at once, and for a detailed changelog. I made a few comments. Very solid work.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js (hunk 1 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:27]: Will do.

File: packages/app/app/components/LibraryView/LibraryFolderTree/styles.scss (hunk 4 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/styles.scss:8]: I'm not sure of the details of how scss variable-insertion works, but I'll try it.

File: packages/app/app/components/LibraryView/LibrarySimpleList/index.js (hunk 5 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibrarySimpleList/index.js:26]: Good point, I'll try that. (I basically only use js-inline styles in my projects, so it didn't come to mind at first)

File: packages/app/app/components/LibraryView/index.js (hunk 6 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/index.js:2]: Right; the action creators and state subscriptions should be moved to the container for consistency.

I got a bit confused, subconsciously thinking the `useSelector` calls were needed in `LibraryView`, since I extracted them from the `useMemo` calls (which do need to be in `LibraryView`). I'll move them to the container.

File: packages/app/app/store/enhancers/syncStorage.js (hunk 7 above)
Comment (Venryx) [reply packages/app/app/store/enhancers/syncStorage.js:11]: From what I remember, without the merging of `reducerInitialStates` beforehand, the supplying of data for `state.local` within the `persistedState` object, was causing the "default value" for the `state` parameter in the `local` reducer to be ignored. (since the reducer function's default-value is only used when `undefined` is passed for that `state` parameter)

I'll double-check this though.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js (hunk 3 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:164]: Will do.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js (hunk 1 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:27]: Will do.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js (hunk 2 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:26]: Will do.

Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:26]: Hmm, but making this change would require that the track metadata be created differently per platform. Isn't it best to standardize on "/" within the persistent metadata, rather than change it per platform? For example, using path.sep for the `track.path` property would mean you couldn't directly transfer the metadata between different-os computers.

File: packages/app/app/components/LibraryView/LibraryFolderTree/styles.scss (hunk 4 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/styles.scss:8]: Okay yeah, I tried placing the import at the top, and it makes it ignore the variable assignments. (the variables have to be set before the import, since the imported scss makes use of those variables, and the variable assignments and reads apparently "run" sequentially)

File: packages/app/app/components/LibraryView/LibrarySimpleList/index.js (hunk 5 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibrarySimpleList/index.js:26]: There doesn't seem to be any way to "identify" the particular element in question, other than by using the `ref` property. Here is the react-list source: https://github.com/coderiety/react-list/blob/master/react-list.es6

Note that there is no `className` or generic props passthrough for the root container. (which is what we need to style)

I could add the style under `.tracks_container div:nth-child(2)`, except then it would apply the styling to the album and folder-tree views as well.

Comment (Venryx) [reply packages/app/app/components/LibraryView/LibrarySimpleList/index.js:26]: One "hack" is to add an extra container around the react-list component, with a unique class. Then I could add the styling for `.library_simple_list_container > div`.

Should I do that, or stick with the ref approach?

File: packages/app/app/store/enhancers/syncStorage.js (hunk 7 above)
Comment (Venryx) [reply packages/app/app/store/enhancers/syncStorage.js:11]: Yep, the description above is correct; without that `reducerInitialState` merging, the default values for the fields in `state.local` (in `reducers/local.js`) are excluded on program start, due to there being data for that root `state.local` field from the `persistedState` source.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js (hunk 1 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:25]: Fixed.

Comment (Venryx) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js]: Fixed.

File: packages/app/app/components/LibraryView/index.js (hunk 6 above)
Comment (Venryx) [reply packages/app/app/components/LibraryView/index.js]: Fixed.

**Comment by Venryx** [comment]
Added most of the changes requested -- except for two which are awaiting comments/confirmation.

File: packages/app/app/components/LibraryView/LibraryFolderTree/index.js (hunk 2 above)
Comment (nukeop) [reply packages/app/app/components/LibraryView/LibraryFolderTree/index.js:24]: The persisted paths don't need to be uniform across platforms, since the directory structures won't match on different machines anyway (you won't be able to transfer the library data to a machine running a different OS). I think you're right though, let's stick with your initial idea, it's simpler.

File: packages/app/app/components/LibraryView/LibrarySimpleList/index.js (hunk 5 above)
Comment (nukeop) [reply packages/app/app/components/LibraryView/LibrarySimpleList/index.js:26]: Ok, that's too bad that the library doesn't expose a way to get to that container.

You could import the styles from the album view and tree view and use the `:not` selector but that's too much code for little benefit. Instead of directly setting the `display` property using the ref, could you please set a class on the ref and use that class in scss? This way if we want to apply more styling to that component in the future it will be easier.

Comment (Venryx) [reply packages/app/app/components/LibraryView/LibrarySimpleList/index.js:26]: Done.

**Comment by Venryx** [comment]
Okay, finished resolving the last change (regarding simple-list-container styling).

**Comment by nukeop** [comment]
Excellent, thank you so much for contributing this.
